core/src/test/java/org/apache/calcite/test/enumerable/EnumerableRecursiveUnionTest.java

```diff
@@ -0,0 +1,357 @@
+                                Arrays.asList("n", "fact"))
+                        .recursiveUnion("D")
+                        .build()
+            )
+            .returnsOrdered("n=0; fact=1",
+                    "n=1; fact=1",
+                    "n=2; fact=2",
+                    "n=3; fact=6",
+                    "n=4; fact=24",
+                    "n=5; fact=120",
+                    "n=6; fact=720",
+                    "n=7; fact=5040");
+  }
+
+  @Test public void testGenerateNumbersNestedRecursion() {
+    CalciteAssert.that()
+            .query("?")
+            .withRel(
+                    //   WITH RECURSIVE t_out(n) AS (
+                    //     WITH RECURSIVE t_in(n) AS (
+                    //       VALUES (1)
+                    //       UNION ALL
+                    //       SELECT n+1 FROM t_in WHERE n < 9
+                    //     )
+                    //     SELECT n FROM t_in
+                    //     UNION ALL
+                    //     SELECT n*10 FROM t_out WHERE n < 100
+                    //   )
+                    //   SELECT n FROM t_out
+                    builder -> builder
+                        .values(new String[] { "n" }, 1)
+                        .deltaScan("T_IN")
+                    .filter(
+                            builder.call(
+                                    SqlStdOperatorTable.LESS_THAN,
+                                    builder.field("n"),
+                                    builder.literal(9)))
+                    .project(
+                            builder.call(
+                                    SqlStdOperatorTable.PLUS,
+                                    builder.field("n"),
+                                    builder.literal(1)))
+                    .recursiveUnion("T_IN")
+
+                    .deltaScan("T_OUT")
+                    .filter(
+                            builder.call(
+                                    SqlStdOperatorTable.LESS_THAN,
+                                    builder.field("n"),
+                                    builder.literal(100)))
+                    .project(
+                            builder.call(
+                                    SqlStdOperatorTable.MULTIPLY,
+                                    builder.field("n"),
+                                    builder.literal(10)))
+                    .recursiveUnion("T_OUT")
+                    .build()
+            )
+            .returnsOrdered(
+                    "n=1",   "n=2",   "n=3",   "n=4",   "n=5",   "n=6",   "n=7",   "n=8",   "n=9",
+                    "n=10",  "n=20",  "n=30",  "n=40",  "n=50",  "n=60",  "n=70",  "n=80",  "n=90",
+                    "n=100", "n=200", "n=300", "n=400", "n=500", "n=600", "n=700", "n=800", "n=900"
+      );
+  }
+
+  // Tests for the following hierarchy:
+  //      Emp1
+  //      /  \
+  //    Emp2  Emp4
+  //    /  \
+  // Emp3   Emp5
+  @Test public void testHierarchy() {
```

Comment:
The test looks like a good candidate for being parameterized (see classes anotated with @RunWith(Parameterized.class)).

Reply:
+1

Reply:
Refactored